File: CS/Q264421/MyFilterControl.cs
using System;
using System.Drawing;
using System.Reflection;
using DevExpress.XtraEditors;
using DevExpress.Utils.Drawing;
using System.Collections.Generic;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Filtering;

namespace DXSample {
    public class MyFilterControl :FilterControl {
        protected override WinFilterTreeNodeModel CreateModel() {
            return new MyWinFilterTreeNodeModel(this);
        }
    }

    public class MyFilterControlLabelInfo :FilterControlLabelInfo {
        public MyFilterControlLabelInfo(Node node) : base(node) { }

        public override void Paint(ControlGraphicsInfoArgs info) {
            ViewInfo.Calculate(info.Graphics);
            ViewInfo.TopLine = 0;
            for (int i = 0; i < ViewInfo.Count; ++i) {
                FilterLabelInfoTextViewInfo textViewInfo = (FilterLabelInfoTextViewInfo)ViewInfo[i];
                NodeEditableElement nodeElement = textViewInfo.InfoText.Tag as NodeEditableElement;
                ElementType elementType = null == nodeElement ? ElementType.None : nodeElement.ElementType;
                if (elementType == ElementType.Property) {
                    textViewInfo.InfoText.Text = textViewInfo.InfoText.Text.Trim('[', ']');
                }
                ViewInfo.Calculate(info.Graphics);
                ViewInfo[i].Draw(info.Cache, info.ViewInfo.Appearance.GetFont(), ViewInfo[i].InfoText.Color, info.ViewInfo.Appearance.GetStringFormat());
            }
        }
      
    }

    public class MyWinFilterTreeNodeModel :WinFilterTreeNodeModel {
        public MyWinFilterTreeNodeModel(FilterControl control) : base(control) { }

        public override void OnVisualChange(FilterChangedActionInternal action, Node node) {
            if (action == FilterChangedActionInternal.NodeAdded)
                ((Dictionary<Node, FilterControlLabelInfo>)typeof(WinFilterTreeNodeModel).GetField("labels",
                    BindingFlags.Instance | BindingFlags.NonPublic).GetValue(this))[node] =
                    new MyFilterControlLabelInfo(node);
            else if (action == FilterChangedActionInternal.RootNodeReplaced) {
                Dictionary<Node, FilterControlLabelInfo> labels =
                    (Dictionary<Node, FilterControlLabelInfo>)typeof(WinFilterTreeNodeModel).GetField("labels",
                    BindingFlags.Instance | BindingFlags.NonPublic).GetValue(this);
                labels.Clear();
                RecursiveVisitor(RootNode, child => {
                    var info = new MyFilterControlLabelInfo(child);
                    info.Clear();
                    info.CreateLabelInfoTexts();
                    labels[child] = info;
                });
            } else base.OnVisualChange(action, node);
        }
    }
}

File: VB/Q264421/MyFilterControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Reflection
Imports DevExpress.XtraEditors
Imports DevExpress.Utils.Drawing
Imports System.Collections.Generic
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Filtering

Namespace DXSample
	Public Class MyFilterControl
		Inherits FilterControl
		Protected Overrides Function CreateModel() As WinFilterTreeNodeModel
			Return New MyWinFilterTreeNodeModel(Me)
		End Function
	End Class

	Public Class MyFilterControlLabelInfo
		Inherits FilterControlLabelInfo
		Public Sub New(ByVal node As Node)
			MyBase.New(node)
		End Sub

		Public Overrides Sub Paint(ByVal info As ControlGraphicsInfoArgs)
			ViewInfo.Calculate(info.Graphics)
			ViewInfo.TopLine = 0
			For i As Integer = 0 To ViewInfo.Count - 1
				Dim textViewInfo As FilterLabelInfoTextViewInfo = CType(ViewInfo(i), FilterLabelInfoTextViewInfo)
				Dim nodeElement As NodeEditableElement = TryCast(textViewInfo.InfoText.Tag, NodeEditableElement)
                Dim _elementType As ElementType = If(Nothing Is nodeElement, ElementType.None, nodeElement.ElementType)
                If _elementType = ElementType.Property Then
                    textViewInfo.InfoText.Text = textViewInfo.InfoText.Text.Trim("["c, "]"c)
                End If
				ViewInfo.Calculate(info.Graphics)
				ViewInfo(i).Draw(info.Cache, info.ViewInfo.Appearance.GetFont(), ViewInfo(i).InfoText.Color, info.ViewInfo.Appearance.GetStringFormat())
			Next i
		End Sub

	End Class

	Public Class MyWinFilterTreeNodeModel
		Inherits WinFilterTreeNodeModel
		Public Sub New(ByVal control As FilterControl)
			MyBase.New(control)
		End Sub

        Public Overrides Sub OnVisualChange(ByVal _action As FilterChangedActionInternal, ByVal node As Node)
            If _action = FilterChangedActionInternal.NodeAdded Then
                CType(GetType(WinFilterTreeNodeModel).GetField("labels", BindingFlags.Instance Or BindingFlags.NonPublic).GetValue(Me), Dictionary(Of Node, FilterControlLabelInfo))(node) = New MyFilterControlLabelInfo(node)
            ElseIf _action = FilterChangedActionInternal.RootNodeReplaced Then
                Dim labels As Dictionary(Of Node, FilterControlLabelInfo) = CType(GetType(WinFilterTreeNodeModel).GetField("labels", BindingFlags.Instance Or BindingFlags.NonPublic).GetValue(Me), Dictionary(Of Node, FilterControlLabelInfo))
                labels.Clear()
                RecursiveVisitor(RootNode, Function(child)
                                               Dim info As New MyFilterControlLabelInfo(child)
                                               info.Clear()
                                               info.CreateLabelInfoTexts()
                                               labels(child) = info
                                               Return New Object()
                                           End Function)
            Else
                MyBase.OnVisualChange(_action, node)
            End If
        End Sub
	End Class
End Namespace